rhp/v4/rhp.go

```diff
@@ -0,0 +1,754 @@
+}
+
+// RPCModifySectorsCost returns the cost of modifying a contract's sectors with the
+// given actions. The duration parameter is the number of blocks until the
+// contract's expiration height.
+func (hp HostPrices) RPCModifySectorsCost(actions []ModifyAction) (cost types.Currency) {
+	var n uint64
+	for _, action := range actions {
+		if action.Type == ActionTrim {
+			n += action.N
+		} else {
+			n++
+		}
+	}
+	return hp.ModifySectorActionPrice.Mul64(n)
+}
+
+// RPCAppendSectorsCost returns the cost of appending sectors to a contract. The duration
+// parameter is the number of blocks until the contract's expiration height.
+func (hp HostPrices) RPCAppendSectorsCost(sectors, duration uint64) (cost, collateral types.Currency) {
+	storage, collateral := hp.StoreSectorCost(duration)
+	storage = storage.Mul64(sectors)
+	collateral = collateral.Mul64(sectors)
+	return storage, collateral
+}
+
+// SigHash returns the hash of the host settings used for signing.
+func (hp HostPrices) SigHash() types.Hash256 {
+	h := types.NewHasher()
+	types.V2Currency(hp.ContractPrice).EncodeTo(h.E)
+	types.V2Currency(hp.Collateral).EncodeTo(h.E)
+	types.V2Currency(hp.StoragePrice).EncodeTo(h.E)
+	types.V2Currency(hp.IngressPrice).EncodeTo(h.E)
+	types.V2Currency(hp.EgressPrice).EncodeTo(h.E)
+	types.V2Currency(hp.ModifySectorActionPrice).EncodeTo(h.E)
+	h.E.WriteUint64(hp.TipHeight)
+	h.E.WriteTime(hp.ValidUntil)
+	return h.Sum()
+}
+
+// Validate checks the host prices for validity. It returns an error if the
+// prices have expired or the signature is invalid.
+func (hp *HostPrices) Validate(pk types.PublicKey) error {
+	if time.Until(hp.ValidUntil) <= 0 {
+		return ErrPricesExpired
+	}
+	if !pk.VerifyHash(hp.SigHash(), hp.Signature) {
+		return ErrInvalidSignature
+	}
+	return nil
+}
+
+// HostSettings specify the settings of a host.
+type HostSettings struct {
+	ProtocolVersion     [3]uint8       `json:"protocolVersion"`
+	Release             string         `json:"release"`
+	WalletAddress       types.Address  `json:"walletAddress"`
+	AcceptingContracts  bool           `json:"acceptingContracts"`
+	MaxCollateral       types.Currency `json:"maxCollateral"`
+	MaxContractDuration uint64         `json:"maxContractDuration"`
+	MaxSectorDuration   uint64         `json:"maxSectorDuration"`
+	MaxModifyActions    uint64         `json:"maxModifyActions"`
+	RemainingStorage    uint64         `json:"remainingStorage"`
+	TotalStorage        uint64         `json:"totalStorage"`
+	Prices              HostPrices     `json:"prices"`
+}
+
+// A ModifyAction adds or modifies sectors within a contract.
+type ModifyAction struct {
+	Type uint8         `json:"type"`
+	Root types.Hash256 `json:"root,omitempty"` // Append
+	A    uint64        `json:"a,omitempty"`    // Swap
+	B    uint64        `json:"b,omitempty"`    // Swap
+	N    uint64        `json:"n,omitempty"`    // Trim
+}
+
+// WriteAction types.
+const (
+	ActionSwap = iota + 1
+	ActionTrim
+	ActionUpdate
+)
```

Comment:
IIRC you mentioned something about RPCModifySectors supporting fewer action types? Need to know for the Merkle proof code.

Reply:
RPC Modify supports swap, trim, and update. Append is now a standalone RPC. The benefit being that the renter may not be 100% sure a sector is stored on the host when pinning roots to a contract. The renter can batch a large number of roots without concern that it will fail in the middle and have to retry.